pyk/src/pyk/proof/proof.py

```diff
@@ -303,20 +304,110 @@ def lines(self) -> list[str]:
+class FailureInfo: ...
 
 
-class StepResult: ...
+class ProverPool(ContextManager['ProverPool'], Generic[P, PS, SR]):
+    _create_prover: Callable[[], Prover[P, PS, SR]]
+    _provers: dict[str, Prover[P, PS, SR]]
+    _executor: Executor
+    _closed: bool
 
+    def __init__(
+        self,
+        create_prover: Callable[[], Prover[P, PS, SR]],
+        *,
+        max_workers: int | None = None,
+    ) -> None:
+        self._create_prover = create_prover
+        self._provers = {}
+        self._executor = ThreadPoolExecutor(max_workers)
+        self._closed = False
 
-class FailureInfo: ...
+    def __enter__(self) -> ProverPool[P, PS, SR]:
+        self._executor.__enter__()
+        return self
 
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self._executor.__exit__(exc_type, exc_val, exc_tb)
+        self.close()
 
-class Prover(Generic[P, PS, SR]):
-    kcfg_explore: KCFGExplore
+    def close(self) -> None:
+        self._closed = True
+        for prover in self._provers.values():
+            prover.shutdown()
+
+    def submit(self, proof_step: PS) -> Future[Iterable[SR]]:
+        if self._closed:
+            raise ValueError('ProverPool has been closed')
+        return self._executor.submit(self._with_prover(proof_step))
+
+    def _with_prover(self, proof_step: PS) -> Callable[[], Iterable[SR]]:
+
+        def step() -> Iterable[SR]:
+            thread_name = current_thread().name
+            prover: Prover[P, PS, SR] | None = self._provers.get(thread_name)
+            if prover is None:
+                prover = self._create_prover()
+                self._provers[thread_name] = prover
+            return prover.step_proof(proof_step)
+
+        return step
+
+
+def parallel_advance_proof(
+    proof: P,
+    create_prover: Callable[[], Prover[P, PS, SR]],
+    max_iterations: int | None = None,
+    fail_fast: bool = False,
+    max_workers: int = 1,
+) -> None:
+    pending: set[Future[Any]] = set()
+    explored: set[PS] = set()
+    iterations = 0
+
+    main_prover = create_prover()
 
-    def __init__(self, kcfg_explore: KCFGExplore):
-        self.kcfg_explore = kcfg_explore
+    main_prover.init_proof(proof)
+
+    with ProverPool[P, PS, SR](create_prover=create_prover, max_workers=max_workers) as pool:
+
+        def submit_steps(_steps: Iterable[PS]) -> None:
+            for step in _steps:
+                if step in explored:
+                    continue
+                explored.add(step)
+                future: Future[Any] = pool.submit(step)  # <-- schedule steps for execution
+                pending.add(future)
+
+        submit_steps(proof.get_steps())
+
+        while True:
+            if len(pending) == 0:
+                break
+            done, _ = wait(pending, return_when='FIRST_COMPLETED')
+            future = done.pop()
+            proof_results = future.result()
+            for result in proof_results:
+                proof.commit(result)
```

Comment:
https://downloads.haskell.org/~ghc/7.4.1/docs/html/users_guide/runtime-control.html#rts-options-environment

Reply:
That's correct. Commit should only be run on the main thread.